### ISSUE TITLE
SyntaxProvider.ForAttributeWithMetadataName skips [field: Attribute] on auto-property, and both that and [property: Attribute] on record primary constructor parameter

### ISSUE THREAD

**Issue by pmg23** [dotnet/roslyn#80511]
**Version Used**: 
Microsoft.CodeAnalysis.Analyzers 3.3.4

**Steps to Reproduce**:

1. In an `IIncrementalGenerator`, call `context.SyntaxProvider.ForAttributeWithMetadataName` with the fully-qualified name of some attribute.
2. Apply the attribute to various elements of the target codebase.
3. Observe that some but not all legal attribute usages are detected.

This expands upon https://github.com/dotnet/roslyn/issues/79575, which reports only that `[method: Attribute]` on primary constructors are missed.

A minimal repro:

```csharp
using Microsoft.CodeAnalysis;

namespace RoslynBugDemo;

[Generator]
internal sealed class Generator : IIncrementalGenerator
{
    public void Initialize(IncrementalGeneratorInitializationContext context)
    {
        var found = context.SyntaxProvider.ForAttributeWithMetadataName("RoslynBugDemo.MarkerAttribute", predicate: static (_, _) => true, transform: (ctx, _) => Diagnostic.Create(
            id: "MYGENERATOR01",
            category: "MyCategory",
            message: $"found {ctx.TargetSymbol.ToDisplayString()}",
            severity: DiagnosticSeverity.Error,
            defaultSeverity: DiagnosticSeverity.Error,
            isEnabledByDefault: true,
            warningLevel: 0,
            location: ctx.TargetNode.GetLocation()
        ));
        context.RegisterSourceOutput(found, (ctx, diagnostic) => ctx.ReportDiagnostic(diagnostic));
    }
}
```

```csharp
using System;

namespace RoslynBugDemo;

internal sealed class MarkerAttribute : Attribute;

[method: Marker] // missing
internal sealed class A(
    [Marker] int B,
    [param: Marker] int C
)
{
    [Marker]
    public int D { get; init; }

    [property: Marker]
    public int E { get; init; }

    [field: Marker] // missing
    public int F { get; init; }

    [Marker]
    public int G() => B + C;

    [return: Marker]
    public int H() => B + C;

    [method: Marker]
    public int I() => B + C;

    [Marker]
    public int J => B + C;

    [property: Marker]
    public int K => B + C;
}

[method: Marker] // missing
internal sealed record L(
    [Marker] int M,
    [property: Marker] int N, // missing
    [field: Marker] int O, // missing
    [param: Marker] int P
)
{
    [Marker]
    public int Q { get; init; }

    [property: Marker]
    public int R { get; init; }

    [field: Marker] // missing
    public int S { get; init; }
}
```

**Expected Behavior**:
Diagnostics reported for each element A through S.

**Actual Behavior**:
```
40>RoslynBugDemo.cs(9,5): Error MYGENERATOR01 : found int B
40>RoslynBugDemo.cs(10,5): Error MYGENERATOR01 : found int C
40>RoslynBugDemo.cs(13,5): Error MYGENERATOR01 : found RoslynBugDemo.A.D
40>RoslynBugDemo.cs(16,5): Error MYGENERATOR01 : found RoslynBugDemo.A.E
40>RoslynBugDemo.cs(22,5): Error MYGENERATOR01 : found RoslynBugDemo.A.G()
40>RoslynBugDemo.cs(25,5): Error MYGENERATOR01 : found RoslynBugDemo.A.H()
40>RoslynBugDemo.cs(28,5): Error MYGENERATOR01 : found RoslynBugDemo.A.I()
40>RoslynBugDemo.cs(31,5): Error MYGENERATOR01 : found RoslynBugDemo.A.J
40>RoslynBugDemo.cs(34,5): Error MYGENERATOR01 : found RoslynBugDemo.A.K
40>RoslynBugDemo.cs(40,5): Error MYGENERATOR01 : found int M
40>RoslynBugDemo.cs(43,5): Error MYGENERATOR01 : found int P
40>RoslynBugDemo.cs(46,5): Error MYGENERATOR01 : found RoslynBugDemo.L.Q
40>RoslynBugDemo.cs(49,5): Error MYGENERATOR01 : found RoslynBugDemo.L.R
```
With missing diagnostics for A, F, L, N, O, S.

**Comment by CyrusNajmabadi** [comment]
As these attributes exist entirely for SG purposes, can you just place them on the nodes where they are found, and then just have arguments tot hem if you want them to apply to the sub symbols?

FAWMN is *very* syntax heavy.  It basically jsut finds matches, then asks the symbol they are directly on if that symbol itself has the attribute the user is looking for.  Complex remappings through targets is not something it was intended for.

IN your case, that could be done with something like `[Marker(forField: true)]`.

**Comment by pmg23** [comment]
Thanks for your reply.

> As these attributes exist entirely for SG purposes ...

The Marker attribute in my example was only for exposition. An actual source generator use case I have finds usages of a custom attribute, declared with `[AttributeUsage(AttributeTargets.Property)]`, that is also used at runtime, i.e. does _not_ exist only for SG. Logically it belongs on the property, and is widely used on properties of regular classes. If the user decides to make such a class into a record with primary constructor syntax, property attributes move equivalently to a `[property: ]` target list on the constructor parameter (my IDE does this code transformation automatically). At runtime (by reflection) there is no change in behaviour, but in SG the attribute disappears from the FAWMN view. It would be hard to justify to the user the need to _remove_ the target specifier in this situation (making it target the parameter instead), and furthermore to expand AttributeUsage to parameters to support this, which would be confusing.

Regarding the current behaviour, FAWMN's [doc](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxvalueprovider.forattributewithmetadataname?view=roslyn-dotnet-4.13.0#microsoft-codeanalysis-syntaxvalueprovider-forattributewithmetadataname-1(system-string-system-func((microsoft-codeanalysis-syntaxnode-system-threading-cancellationtoken-system-boolean))-system-func((microsoft-codeanalysis-generatorattributesyntaxcontext-system-threading-cancellationtoken-0)))) says

> Creates an IncrementalValuesProvider<TValues> that can provide a transform over all SyntaxNodes if that node has an attribute on it that binds to a INamedTypeSymbol with the same fully-qualified metadata as the provided fullyQualifiedMetadataName.

Are you arguing that the current behaviour is correct because there is in fact no syntax node meeting this description? If so I think the doc at least needs clarification.

In my opinion, it _should_ discover all instances of the attribute, wherever the language allows it to appear - now and in the future. (If FAWMN does not guarantee this it becomes practically unusable for certain use cases such as mine, and a pitfall for anyone who might naively assume it.) If there is a concern about performance I suggest FAWMN could take an AttributeTargets (or AttributeLocation) flags enum value that specifies where to look, allowing for redundant code to be skipped.

**Comment by CyrusNajmabadi** [comment]
Yes.  I agree this should likely be fixed.  Just pointing out it is non-trival.  Cheers.

**Comment by Rekkonnect** [comment]
Bumping this thread because I also stumbled upon it. 

Not a massive deal-breaker but definitely undocumented and unexpected lack of support. It feels that it's also quite computationally cheap to fix, i.e. check what the attribute target specifier is, and query the appropriate symbol's attributes instead, if available. Invalid targets will of course be ignored. 

Off-topic, but haven't seen any attribute target helper exposed in the public Roslyn API surface that gives the symbol that's intended to receive the attribute. Not that it's required, but I'd assume it would be equally helpful for more cases that analyze attributes. Just throwing it out in case the idea comes by also implementing the extension of FAWMN for different attribute targets.
